File: nodejs/test/e2e/system_message_transform.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { writeFile } from "fs/promises";
import { join } from "path";
import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("System message transform", async () => {
    const { copilotClient: client, workDir } = await createSdkTestContext();

    it("should invoke transform callbacks with section content", async () => {
        const transformedSections: Record<string, string> = {};

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            systemMessage: {
                mode: "customize",
                sections: {
                    identity: {
                        action: (content: string) => {
                            transformedSections["identity"] = content;
                            // Pass through unchanged
                            return content;
                        },
                    },
                    tone: {
                        action: (content: string) => {
                            transformedSections["tone"] = content;
                            return content;
                        },
                    },
                },
            },
        });

        await writeFile(join(workDir, "test.txt"), "Hello transform!");

        await session.sendAndWait({
            prompt: "Read the contents of test.txt and tell me what it says",
        });

        // Transform callbacks should have been invoked with real section content
        expect(Object.keys(transformedSections).length).toBe(2);
        expect(transformedSections["identity"]).toBeDefined();
        expect(transformedSections["identity"]!.length).toBeGreaterThan(0);
        expect(transformedSections["tone"]).toBeDefined();
        expect(transformedSections["tone"]!.length).toBeGreaterThan(0);

        await session.disconnect();
    });

    it("should apply transform modifications to section content", async () => {
        let originalContent = "";
        let transformedContent = "";

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            systemMessage: {
                mode: "customize",
                sections: {
                    identity: {
                        action: (content: string) => {
                            originalContent = content;
                            // Append a custom instruction via transform
                            transformedContent = content + "\nTRANSFORM_MARKER";
                            return transformedContent;
                        },
                    },
                },
            },
        });

        await writeFile(join(workDir, "hello.txt"), "Hello!");

        await session.sendAndWait({
            prompt: "Read the contents of hello.txt",
        });

        // Verify the transform callback was invoked and modified the content
        expect(originalContent.length).toBeGreaterThan(0);
        expect(transformedContent).toContain("TRANSFORM_MARKER");
        expect(transformedContent).toContain(originalContent);

        await session.disconnect();
    });

    it("should work with static overrides and transforms together", async () => {
        const transformedSections: Record<string, string> = {};

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            systemMessage: {
                mode: "customize",
                sections: {
                    // Static override
                    safety: { action: "remove" },
                    // Transform
                    identity: {
                        action: (content: string) => {
                            transformedSections["identity"] = content;
                            return content;
                        },
                    },
                },
            },
        });

        await writeFile(join(workDir, "combo.txt"), "Combo test!");

        await session.sendAndWait({
            prompt: "Read the contents of combo.txt and tell me what it says",
        });

        // Transform should have been invoked
        expect(transformedSections["identity"]).toBeDefined();
        expect(transformedSections["identity"]!.length).toBeGreaterThan(0);

        await session.disconnect();
    });
});

File: go/internal/e2e/system_message_transform_test.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package e2e

import (
	"os"
	"path/filepath"
	"sync"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestSystemMessageTransform(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_invoke_transform_callbacks_with_section_content", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		var identityContent string
		var toneContent string
		var mu sync.Mutex
		identityCalled := false
		toneCalled := false

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SystemMessage: &copilot.SystemMessageConfig{
				Mode: "customize",
				Sections: map[string]copilot.SectionOverride{
					"identity": {
						Transform: func(currentContent string) (string, error) {
							mu.Lock()
							identityCalled = true
							identityContent = currentContent
							mu.Unlock()
							return currentContent, nil
						},
					},
					"tone": {
						Transform: func(currentContent string) (string, error) {
							mu.Lock()
							toneCalled = true
							toneContent = currentContent
							mu.Unlock()
							return currentContent, nil
						},
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		testFile := filepath.Join(ctx.WorkDir, "test.txt")
		err = os.WriteFile(testFile, []byte("Hello transform!"), 0644)
		if err != nil {
			t.Fatalf("Failed to write test file: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the contents of test.txt and tell me what it says",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		mu.Lock()
		defer mu.Unlock()

		if !identityCalled {
			t.Error("Expected identity transform callback to be invoked")
		}
		if !toneCalled {
			t.Error("Expected tone transform callback to be invoked")
		}
		if identityContent == "" {
			t.Error("Expected identity transform to receive non-empty content")
		}
		if toneContent == "" {
			t.Error("Expected tone transform to receive non-empty content")
		}
	})

	t.Run("should_apply_transform_modifications_to_section_content", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		var receivedContent string
		var mu sync.Mutex
		transformCalled := false

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SystemMessage: &copilot.SystemMessageConfig{
				Mode: "customize",
				Sections: map[string]copilot.SectionOverride{
					"identity": {
						Transform: func(currentContent string) (string, error) {
							mu.Lock()
							transformCalled = true
							receivedContent = currentContent
							mu.Unlock()
							return currentContent + "\nTRANSFORM_MARKER", nil
						},
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		testFile := filepath.Join(ctx.WorkDir, "hello.txt")
		err = os.WriteFile(testFile, []byte("Hello!"), 0644)
		if err != nil {
			t.Fatalf("Failed to write test file: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the contents of hello.txt",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		mu.Lock()
		defer mu.Unlock()

		if !transformCalled {
			t.Error("Expected transform callback to be invoked")
		}
		if receivedContent == "" {
			t.Error("Expected transform to receive non-empty content")
		}
	})

	t.Run("should_work_with_static_overrides_and_transforms_together", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		var mu sync.Mutex
		transformCalled := false

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SystemMessage: &copilot.SystemMessageConfig{
				Mode: "customize",
				Sections: map[string]copilot.SectionOverride{
					"safety": {
						Action: copilot.SectionActionRemove,
					},
					"identity": {
						Transform: func(currentContent string) (string, error) {
							mu.Lock()
							transformCalled = true
							mu.Unlock()
							return currentContent, nil
						},
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		testFile := filepath.Join(ctx.WorkDir, "combo.txt")
		err = os.WriteFile(testFile, []byte("Combo test!"), 0644)
		if err != nil {
			t.Fatalf("Failed to write test file: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the contents of combo.txt and tell me what it says",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		mu.Lock()
		defer mu.Unlock()

		if !transformCalled {
			t.Error("Expected identity transform callback to be invoked")
		}
	})
}
